modules/fb_instant_articles_display/src/EntityPropertyMapper.php

```diff
@@ -0,0 +1,209 @@
+use Facebook\InstantArticles\Elements\InstantArticle;
+use Facebook\InstantArticles\Elements\Time;
+
+/**
+ * Class EntityPropertyMapper
+ * @package Drupal\fb_instant_articles_display
+ */
+class EntityPropertyMapper {
+
+  /**
+   * Entity object for which we are building an InstantArticle object.
+   *
+   * @var \stdClass
+   */
+  private $entity;
+  /**
+   * The Entity type.
+   *
+   * We store this because there is no core $entity->entity_type property.
+   * See @link https://www.drupal.org/node/1042822 this d.o issue for status. @endlink
+   *
+   * @see $entity
+   *
+   * @var string
+   */
+  private $entity_type;
+
+  /**
+   * Facebook Instant Articles object that encapsulates the structure and
+   * content of the Entity object we are wrapping.
+   *
+   * @var InstantArticle
+   */
+  private $instantArticle;
+
+  /**
+   * EntityPropertyMapper constructor.
+   *
+   * @param string $entity_type
+   * @param \stdClass|mixed $entity
+   * @param \Facebook\InstantArticles\Elements\InstantArticle $instantArticle
+   */
+  private function __construct($entity_type, $entity, InstantArticle $instantArticle) {
+    $this->entity_type = $entity_type;
+    $this->entity = $entity;
+    $this->instantArticle = $instantArticle;
+  }
+
+  /**
+   * @param string $entity_type
+   * @param \stdClass $entity
+   * @param InstantArticle $instantArticle
+   * @return EntityPropertyMapper
+   */
+  public static function create($entity_type, $entity, InstantArticle $instantArticle) {
+    return new EntityPropertyMapper($entity_type, $entity, $instantArticle);
+  }
+
+  /**
+   * Maps Drupal Entity properties to a FB Instant Article SDK InstantArticle.
+   */
+  public function map() {
+    $this->addCanonicalURL();
+    $this->addHeaderFromProperties();
+    $this->addAnalyticsFromSettings();
+    $this->addAdsFromSettings();
+  }
+
+  private function addCanonicalURL() {
+    $path = entity_uri($this->entity_type, $this->entity);
+    $this->instantArticle->withCanonicalUrl(url($path['path'], array('absolute' => TRUE)));
+  }
+
+  private function addHeaderFromProperties() {
+    $header = Header::create();
+    if ($label = entity_label($this->entity_type, $this->entity)) {
+      $header->withTitle($label);
+    }
+
+    // Support specific Drupal Entity keys to map to the Instant SDK PUBLISHED
+    // and MODIFIED concepts.
+    if (isset($this->entity->created)) {
+      $header->withPublishTime(
+        Time::create(Time::PUBLISHED)
+          ->withDatetime(
+            \DateTime::createFromFormat('U', $this->entity->created)
+          )
+      );
+    }
+    if (isset($this->entity->changed)) {
+      $header->withModifyTime(
+        Time::create(Time::MODIFIED)
+          ->withDatetime(
+            \DateTime::createFromFormat('U', $this->entity->changed)
+          )
+      );
+    }
+
+    // Default the article author to the username.
+    if (isset($this->entity->uid)) {
+      $author = user_load($this->entity->uid);
```

Comment:
With the change to running `hook_fb_instant_articles_article_alter` on create, I wonder if that will make the author field on the instant article impossible to override?  Since this will come after the alter has run and set it.  Will test.